### PR TITLE
Insertion markers and replacement markers for value inputs and termin…

### LINKED ISSUE #370: Design - what happens when a block is dragged over an input?
In Scratch 2.0, a white "glow" appears around the old text area/block.
In Blockly, the yellow notch is highlighted.
In Scratch Blocks currently, there are two cases:
- If the input is occupied by a shadow block, an insertion marker replaces the shadow.
- If the input is occupied by another block, the other block is bumped out of the position and replaced by the insertion marker. This appears as a bug, because the other block is never replaced if the drag moves somewhere else.

Some options that seem viable:
- Scratch 2.0 behavior for all cases (whether or not there's a shadow). If we do this, it would be great to have a spec for size/color/blur/etc. of a stroke.
- Bump any block in an input, but replace it back in when the drag moves somewhere else.
- Some mix of the two: e.g., use the insertion marker for a shadow block, but some Scratch 2.0-style stroke for regular blocks in inputs.

### LINKED ISSUE #390: block_svg.js:1168 Uncaught insertionMarkerConnection still connected at the end of disconnectInsertionMarker
Sprinkles seems to trigger this. Just logging the issue so we don't forget :)

### LINKED ISSUE #397: Displacing terminal statement blocks
I'm looking at Blockly [Issue 414](https://github.com/google/blockly/issues/414).  Although that particular issue can be resolved by special casing shadow blocks, the root cause is a change in behaviour made to Blockly to match Scratch.  I'm questioning whether Scratch's behaviour is desirable.

Here is a movie showing Scratch's restricted options when dealing with terminal statement blocks.
  https://youtu.be/Qt33f_J6YxU
Note that this is inconsistent with all other block types, where one can displace an existing block by dragging in a replacement.

If a user drags a terminal statement block over another terminal statement block, it is clear what they want to do.  Can we remove the special exception we added that prevents the user from accomplishing their intended goal?

### PR DESCRIPTION
**Pull request by rachel-fenichel**
…al blocks

Fixes #397 
Fixes #390 
Mebbe fixes #370?

### CODE REVIEW

**Comment by tmickel** [comment]
Wow! I'm really excited at how well this works!

File: core/block_svg.js [hunk 1]
```diff
@@ -1131,6 +1118,81 @@ Blockly.BlockSvg.prototype.updatePreviews = function(closestConnection,
...
+  }
+  Blockly.replacementMarker_ = null;
+};
+
+/**
+ * Place and render an insertion marker to indicate what would happen if you
+ * release the drag right now.
+ * @param {Blockly.Connection} localConnection The connection on the dragging
+ *     block.
+ * @param {Blockly.Connection} closestConnection The connnection to pretend to
```
Comment (tmickel) [inline core/block_svg.js]: I'm a little confused by "pretend to connect to" - the insertion marker is actually connected, right? (but the replacement marker doesn't actually make a connection - "pretend" makes sense to me there).

File: core/block_svg.js [hunk 2]
```diff
@@ -1085,39 +1089,22 @@ Blockly.BlockSvg.prototype.updatePreviews = function(closestConnection,
...
-          insertionMarkerConnection, closestConnection);
-
-      if (insertionMarkerConnection.type == Blockly.PREVIOUS_STATEMENT &&
-          !insertionMarker.nextConnection) {
-        Blockly.bumpedConnection_ = closestConnection.targetConnection;
-      }
-      // Renders insertion marker.
-      insertionMarkerConnection.connect(closestConnection);
-      Blockly.insertionMarkerConnection_ = insertionMarkerConnection;
+    var shouldReplace = (localConnection.type == Blockly.OUTPUT_VALUE ||
```
Comment (tmickel) [inline core/block_svg.js:1102]: Don't strongly care, but might be better to move this logic into its own helper or at least add a comment about the situations where shouldReplace is true.

File: core/connection.js [hunk 3]
```diff
@@ -439,12 +439,8 @@ Blockly.Connection.prototype.isConnectionAllowed = function(candidate) {
...
     }
     case Blockly.OUTPUT_VALUE: {
-      // Don't offer to connect an already connected left (male) value plug to
-      // an available right (female) value plug.
-      if (candidate.targetConnection || this.targetConnection) {
-        return false;
-      }
-      break;
+      // Can't drag an input to an output--you have to move the inferior block.
+      return false;
```
Comment (tmickel) [inline core/connection.js:443]: Glad we fixed this as well. Blockly is going to be left as-is for this case, right?

File: core/block_svg.js (hunk 1 above)
Comment (rachel-fenichel) [inline core/block_svg.js]: How about "The connection to connect the insertion marker to"?

Split infinitives be damned.

File: core/connection.js [hunk 4]
```diff
@@ -469,9 +465,13 @@ Blockly.Connection.prototype.isConnectionAllowed = function(candidate) {
...
       // Don't let a block with no next connection bump other blocks out of the
-      // stack.
+      // stack.  But covering up a shadow block or stack of shadow blocks is
+      // fine.  Similarly, replacing a terminal statement with another terminal
+      // statement is allowed.
       if (candidate.isConnectedToNonInsertionMarker() &&
-          !this.sourceBlock_.nextConnection) {
+          !this.sourceBlock_.nextConnection &&
+          !candidate.targetBlock().isShadow() &&
+          candidate.targetBlock().nextConnection) {
```
Comment (tmickel) [inline core/connection.js:474]: Looks good. I like how it works, too, even with stacks that ends in terminal blocks (I think it's fine that it bumps and allows this situation).

**Comment by tmickel** [comment]
Thanks again - LGTM

File: core/connection.js (hunk 3 above)
Comment (rachel-fenichel) [inline core/connection.js:443]: Yeah, blockly absolutely allows this.  I was surprised that Scratch doesn't--but this is what was causing a lot of the problems on the vertical playground.

File: core/block_svg.js (hunk 2 above)
Comment (rachel-fenichel) [inline core/block_svg.js:1102]: I'll add a comment.

File: core/block_svg.js (hunk 1 above)
Comment (tmickel) [inline core/block_svg.js]: ^ sounds better :)

**Pull request merged**